File: python/src/wheelx_sdk/wheelx_sdk.py
"""
WheelX Python SDK for quote API and transaction execution
"""

import json
import time
from typing import Dict, Optional, Any
from dataclasses import dataclass
import requests


@dataclass
class QuoteRequest:
    """Quote request parameters"""
    from_chain: int
    to_chain: int
    from_token: str
    to_token: str
    from_address: str
    to_address: str
    amount: int
    slippage: Optional[int] = None


@dataclass
class Tx:
    """Transaction data"""
    to: str
    value: int
    data: str
    chainId: Optional[int] = None
    gas: Optional[int] = None
    maxFeePerGas: Optional[int] = None
    maxPriorityFeePerGas: Optional[int] = None


@dataclass
class ApproveAction:
    """Approval transaction data"""
    token: str
    spender: str
    amount: int


@dataclass
class PriceImpactFormatted:
    """Price impact breakdown"""
    bridge_fee: str
    swap_fee: str
    dst_gas_fee: str


@dataclass
class QuoteResponse:
    """Quote response data"""
    request_id: str
    amount_out: str
    fee: str
    tx: Tx
    approve: Optional[ApproveAction]
    slippage: int
    min_receive: str
    estimated_time: int
    recipient: str
    router_type: str
    price_impact: PriceImpactFormatted
    router: str
    created_at: str
    points: str


class WheelXSDK:
    """WheelX SDK for quote API and transaction execution"""

    def __init__(self, base_url: str = "https://wheelx.fi"):
        self.base_url = base_url
        self.session = requests.Session()

    def get_quote(self, quote_request: QuoteRequest) -> QuoteResponse:
        """
        Get a quote for token swap/bridge

        Args:
            quote_request: Quote request parameters

        Returns:
            QuoteResponse: Quote response with transaction data

        Raises:
            Exception: If API request fails
        """
        url = f"{self.base_url}/v1/quote"

        payload = {
            "from_chain": quote_request.from_chain,
            "to_chain": quote_request.to_chain,
            "from_token": quote_request.from_token,
            "to_token": quote_request.to_token,
            "from_address": quote_request.from_address,
            "to_address": quote_request.to_address,
            "amount": quote_request.amount,
        }

        if quote_request.slippage is not None:
            payload["slippage"] = quote_request.slippage

        response = self.session.post(url, json=payload)
        response.raise_for_status()

        data = response.json()

        # Parse transaction data
        tx_data = data.get("tx", {})
        tx = Tx(
            to=tx_data.get("to"),
            value=tx_data.get("value", 0),
            data=tx_data.get("data"),
            chainId=tx_data.get("chainId"),
            gas=tx_data.get("gas"),
            maxFeePerGas=tx_data.get("maxFeePerGas"),
            maxPriorityFeePerGas=tx_data.get("maxPriorityFeePerGas")
        )

        # Parse approval data if present
        approve_data = data.get("approve")
        approve = None
        if approve_data:
            approve = ApproveAction(
                token=approve_data.get("token"),
                spender=approve_data.get("spender"),
                amount=approve_data.get("amount")
            )

        # Parse price impact
        price_impact_data = data.get("price_impact", {})
        price_impact = PriceImpactFormatted(
            bridge_fee=price_impact_data.get("bridge_fee", "0"),
            swap_fee=price_impact_data.get("swap_fee", "0"),
            dst_gas_fee=price_impact_data.get("dst_gas_fee", "0")
        )

        return QuoteResponse(
            request_id=data.get("request_id"),
            amount_out=data.get("amount_out"),
            fee=data.get("fee"),
            tx=tx,
            approve=approve,
            slippage=data.get("slippage"),
            min_receive=data.get("min_receive"),
            estimated_time=data.get("estimated_time", 10),
            recipient=data.get("recipient"),
            router_type=data.get("router_type", "swap"),
            price_impact=price_impact,
            router=data.get("router", "wheelx"),
            created_at=data.get("created_at"),
            points=data.get("points", "0")
        )

    def get_order_status(self, request_id: str) -> Dict[str, Any]:
        """
        Get order status by request ID

        Args:
            request_id: Quote request ID

        Returns:
            Dict: Order status information
        """
        url = f"{self.base_url}/v1/order/{request_id}"
        response = self.session.get(url)
        response.raise_for_status()
        return response.json()


class TransactionExecutor:
    """Helper class for executing transactions using web3.py"""

    def __init__(self, web3_provider_url: str):
        """
        Initialize transaction executor

        Args:
            web3_provider_url: Web3 provider URL (e.g., Infura, Alchemy)
        """
        try:
            from web3 import Web3
            self.web3 = Web3(Web3.HTTPProvider(web3_provider_url))
            if not self.web3.is_connected():
                raise Exception("Failed to connect to Web3 provider")
        except ImportError:
            raise ImportError("web3.py is required for transaction execution. Install with: pip install web3")

    def build_transaction(self, tx_data: Tx, account_address: str) -> Dict[str, Any]:
        """
        Build transaction dictionary for signing

        Args:
            tx_data: Transaction data from quote response
            account_address: User's wallet address

        Returns:
            Dict: Transaction dictionary ready for signing
        """
        transaction = {
            'to': self.web3.to_checksum_address(tx_data.to),
            'value': tx_data.value,
            'data': tx_data.data,
            'chainId': tx_data.chainId,
            'nonce': self.web3.eth.get_transaction_count(account_address),
        }

        # Add gas parameters
        if tx_data.gas:
            transaction['gas'] = tx_data.gas

        # Add EIP-1559 gas parameters
        if tx_data.maxFeePerGas and tx_data.maxPriorityFeePerGas:
            transaction['maxFeePerGas'] = tx_data.maxFeePerGas
            transaction['maxPriorityFeePerGas'] = tx_data.maxPriorityFeePerGas
        else:
            # Fallback to legacy gas
            gas_price = self.web3.eth.gas_price
            transaction['gasPrice'] = gas_price

        return transaction

    def sign_and_send_transaction(self, transaction: Dict[str, Any], private_key: str) -> str:
        """
        Sign and send transaction

        Args:
            transaction: Transaction dictionary
            private_key: Private key for signing

        Returns:
            str: Transaction hash

        Raises:
            Exception: If transaction fails
        """
        signed_txn = self.web3.eth.account.sign_transaction(transaction, private_key)
        tx_hash = self.web3.eth.send_raw_transaction(signed_txn.rawTransaction)
        return self.web3.to_hex(tx_hash)

    def wait_for_transaction(self, tx_hash: str, timeout: int = 300) -> Dict[str, Any]:
        """
        Wait for transaction confirmation

        Args:
            tx_hash: Transaction hash
            timeout: Timeout in seconds

        Returns:
            Dict: Transaction receipt
        """
        return self.web3.eth.wait_for_transaction_receipt(tx_hash, timeout=timeout)


# Example usage
if __name__ == "__main__":
    # Initialize SDK
    sdk = WheelXSDK()

    # Create quote request
    quote_request = QuoteRequest(
        from_chain=1,
        to_chain=1,
        from_token="0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",  # USDC
        to_token="0xdAC17F958D2ee523a2206206994597C13D831ec7",    # USDT
        from_address="0x742d35Cc6634C0532925a3b8Dc9F6A7c5D3a7C6a",
        to_address="0x742d35Cc6634C0532925a3b8Dc9F6A7c5D3a7C6a",
        amount=1000000,  # 1 USDC (6 decimals)
        slippage=50
    )

    try:
        # Get quote
        quote = sdk.get_quote(quote_request)
        print(f"Quote received: {quote.amount_out} tokens")
        print(f"Request ID: {quote.request_id}")
        print(f"Min receive: {quote.min_receive}")

        # Check if approval is needed
        if quote.approve:
            print(f"Approval needed for token: {quote.approve.token}")
            print(f"Spender: {quote.approve.spender}")
            print(f"Amount: {quote.approve.amount}")

        # Example transaction execution (requires web3 and private key)
        # executor = TransactionExecutor("https://mainnet.infura.io/v3/YOUR_PROJECT_ID")
        # transaction = executor.build_transaction(quote.tx, quote_request.from_address)
        # tx_hash = executor.sign_and_send_transaction(transaction, "YOUR_PRIVATE_KEY")
        # print(f"Transaction sent: {tx_hash}")
        # receipt = executor.wait_for_transaction(tx_hash)
        # print(f"Transaction confirmed in block: {receipt.blockNumber}")

    except Exception as e:
        print(f"Error: {e}")

File: go/wheelx/wheelx_sdk.go
package wheelx

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"time"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethclient"
)

// QuoteRequest represents the request parameters for getting a quote
type QuoteRequest struct {
	FromChain   int    `json:"from_chain"`
	ToChain     int    `json:"to_chain"`
	FromToken   string `json:"from_token"`
	ToToken     string `json:"to_token"`
	FromAddress string `json:"from_address"`
	ToAddress   string `json:"to_address"`
	Amount      int    `json:"amount"`
	Slippage    *int   `json:"slippage,omitempty"`
}

// Tx represents transaction data
type Tx struct {
	To                   string `json:"to"`
	Value                int    `json:"value"`
	Data                 string `json:"data"`
	ChainId              *int   `json:"chainId,omitempty"`
	Gas                  *int   `json:"gas,omitempty"`
	MaxFeePerGas         *int   `json:"maxFeePerGas,omitempty"`
	MaxPriorityFeePerGas *int   `json:"maxPriorityFeePerGas,omitempty"`
}

// ApproveAction represents approval transaction data
type ApproveAction struct {
	Token   string `json:"token"`
	Spender string `json:"spender"`
	Amount  int    `json:"amount"`
}

// PriceImpactFormatted represents price impact breakdown
type PriceImpactFormatted struct {
	BridgeFee string `json:"bridge_fee"`
	SwapFee   string `json:"swap_fee"`
	DstGasFee string `json:"dst_gas_fee"`
}

// QuoteResponse represents the quote response
type QuoteResponse struct {
	RequestId     string               `json:"request_id"`
	AmountOut     string               `json:"amount_out"`
	Fee           string               `json:"fee"`
	Tx            Tx                   `json:"tx"`
	Approve       *ApproveAction       `json:"approve,omitempty"`
	Slippage      int                  `json:"slippage"`
	MinReceive    string               `json:"min_receive"`
	EstimatedTime int                  `json:"estimated_time"`
	Recipient     string               `json:"recipient"`
	RouterType    string               `json:"router_type"`
	PriceImpact   PriceImpactFormatted `json:"price_impact"`
	Router        string               `json:"router"`
	CreatedAt     string               `json:"created_at"`
	Points        string               `json:"points"`
}

// OrderResponse represents order status response
type OrderResponse struct {
	OrderId       string `json:"order_id"`
	FromChain     int    `json:"from_chain"`
	FromToken     string `json:"from_token"`
	FromAddress   string `json:"from_address"`
	FromAmount    string `json:"from_amount"`
	ToChain       int    `json:"to_chain"`
	ToToken       string `json:"to_token"`
	ToAmount      string `json:"to_amount"`
	ToAddress     string `json:"to_address"`
	OpenTxHash    string `json:"open_tx_hash"`
	OpenBlock     int    `json:"open_block"`
	OpenTimestamp string `json:"open_timestamp"`
	Status        string `json:"status"`
	Points        string `json:"points"`
}

// WheelXSDK provides methods to interact with WheelX API
type WheelXSDK struct {
	BaseURL string
	Client  *http.Client
}

// NewWheelXSDK creates a new WheelX SDK instance
func NewWheelXSDK(baseURL string) *WheelXSDK {
	if baseURL == "" {
		baseURL = "https://wheelx.fi"
	}

	return &WheelXSDK{
		BaseURL: baseURL,
		Client:  &http.Client{Timeout: 30 * time.Second},
	}
}

// GetQuote gets a quote for token swap/bridge
func (sdk *WheelXSDK) GetQuote(ctx context.Context, req QuoteRequest) (*QuoteResponse, error) {
	url := sdk.BaseURL + "/v1/quote"

	// Prepare request body
	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create HTTP request
	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")

	// Send request
	resp, err := sdk.Client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// Read response
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Check status code
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API request failed: %s - %s", resp.Status, string(respBody))
	}

	// Parse response
	var quoteResp QuoteResponse
	if err := json.Unmarshal(respBody, &quoteResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &quoteResp, nil
}

// GetOrderStatus gets order status by request ID
func (sdk *WheelXSDK) GetOrderStatus(ctx context.Context, requestID string) (*OrderResponse, error) {
	url := sdk.BaseURL + "/v1/order/" + requestID

	httpReq, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := sdk.Client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API request failed: %s - %s", resp.Status, string(respBody))
	}

	var orderResp OrderResponse
	if err := json.Unmarshal(respBody, &orderResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &orderResp, nil
}

// TransactionExecutor provides methods to execute transactions
type TransactionExecutor struct {
	Client *ethclient.Client
}

// NewTransactionExecutor creates a new transaction executor
func NewTransactionExecutor(rpcURL string) (*TransactionExecutor, error) {
	client, err := ethclient.Dial(rpcURL)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to Ethereum client: %w", err)
	}

	return &TransactionExecutor{
		Client: client,
	}, nil
}

// BuildTransaction builds a transaction for signing
func (e *TransactionExecutor) BuildTransaction(ctx context.Context, txData Tx, fromAddress common.Address) (*types.Transaction, error) {
	// Get nonce
	nonce, err := e.Client.PendingNonceAt(ctx, fromAddress)
	if err != nil {
		return nil, fmt.Errorf("failed to get nonce: %w", err)
	}

	// Get gas price
	gasPrice, err := e.Client.SuggestGasPrice(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get gas price: %w", err)
	}

	// Parse transaction data
	data := common.FromHex(txData.Data)
	to := common.HexToAddress(txData.To)

	// Create transaction
	tx := types.NewTx(&types.LegacyTx{
		Nonce:    nonce,
		To:       &to,
		Value:    big.NewInt(int64(txData.Value)),
		Gas:      uint64(*txData.Gas),
		GasPrice: gasPrice,
		Data:     data,
	})

	return tx, nil
}

// BuildEIP1559Transaction builds an EIP-1559 transaction
func (e *TransactionExecutor) BuildEIP1559Transaction(ctx context.Context, txData Tx, fromAddress common.Address) (*types.Transaction, error) {
	// Get nonce
	nonce, err := e.Client.PendingNonceAt(ctx, fromAddress)
	if err != nil {
		return nil, fmt.Errorf("failed to get nonce: %w", err)
	}

	// Get chain ID
	chainID, err := e.Client.ChainID(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get chain ID: %w", err)
	}

	// Get gas tip cap (priority fee)
	gasTipCap, err := e.Client.SuggestGasTipCap(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get gas tip cap: %w", err)
	}

	// Get gas fee cap
	header, err := e.Client.HeaderByNumber(ctx, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get header: %w", err)
	}
	baseFee := new(big.Int).Mul(header.BaseFee, big.NewInt(2))
	gasFeeCap := new(big.Int).Add(gasTipCap, baseFee)

	// Parse transaction data
	data := common.FromHex(txData.Data)
	to := common.HexToAddress(txData.To)

	// Create EIP-1559 transaction
	tx := types.NewTx(&types.DynamicFeeTx{
		ChainID:   chainID,
		Nonce:     nonce,
		GasTipCap: gasTipCap,
		GasFeeCap: gasFeeCap,
		Gas:       uint64(*txData.Gas),
		To:        &to,
		Value:     big.NewInt(int64(txData.Value)),
		Data:      data,
	})

	return tx, nil
}

// SignAndSendTransaction signs and sends a transaction
func (e *TransactionExecutor) SignAndSendTransaction(ctx context.Context, tx *types.Transaction, privateKey string) (common.Hash, error) {
	// Parse private key
	key, err := crypto.HexToECDSA(privateKey)
	if err != nil {
		return common.Hash{}, fmt.Errorf("failed to parse private key: %w", err)
	}

	// Get chain ID
	chainID, err := e.Client.ChainID(ctx)
	if err != nil {
		return common.Hash{}, fmt.Errorf("failed to get chain ID: %w", err)
	}

	// Sign transaction
	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(chainID), key)
	if err != nil {
		return common.Hash{}, fmt.Errorf("failed to sign transaction: %w", err)
	}

	// Send transaction
	err = e.Client.SendTransaction(ctx, signedTx)
	if err != nil {
		return common.Hash{}, fmt.Errorf("failed to send transaction: %w", err)
	}

	return signedTx.Hash(), nil
}

// WaitForTransaction waits for transaction confirmation
func (e *TransactionExecutor) WaitForTransaction(ctx context.Context, txHash common.Hash) (*types.Receipt, error) {
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-ticker.C:
			receipt, err := e.Client.TransactionReceipt(ctx, txHash)
			if err != nil {
				if err.Error() == "not found" {
					continue
				}
				return nil, fmt.Errorf("failed to get transaction receipt: %w", err)
			}
			return receipt, nil
		}
	}
}

// Example usage
/*
func main() {
	// Initialize SDK
	sdk := NewWheelXSDK("")

	// Create quote request
	req := QuoteRequest{
		FromChain:   1,
		ToChain:     1,
		FromToken:   "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48", // USDC
		ToToken:     "0xdAC17F958D2ee523a2206206994597C13D831ec7",   // USDT
		FromAddress: "0x742d35Cc6634C0532925a3b8Dc9F6A7c5D3a7C6a",
		ToAddress:   "0x742d35Cc6634C0532925a3b8Dc9F6A7c5D3a7C6a",
		Amount:      1000000, // 1 USDC (6 decimals)
		Slippage:    &[]int{50}[0],
	}

	// Get quote
	ctx := context.Background()
	quote, err := sdk.GetQuote(ctx, req)
	if err != nil {
		log.Fatalf("Failed to get quote: %v", err)
	}

	fmt.Printf("Quote received: %s tokens\n", quote.AmountOut)
	fmt.Printf("Request ID: %s\n", quote.RequestId)
	fmt.Printf("Min receive: %s\n", quote.MinReceive)

	// Check if approval is needed
	if quote.Approve != nil {
		fmt.Printf("Approval needed for token: %s\n", quote.Approve.Token)
		fmt.Printf("Spender: %s\n", quote.Approve.Spender)
		fmt.Printf("Amount: %d\n", quote.Approve.Amount)
	}

	// Example transaction execution (requires Ethereum client and private key)
	// executor, err := NewTransactionExecutor("https://mainnet.infura.io/v3/YOUR_PROJECT_ID")
	// if err != nil {
	//     log.Fatalf("Failed to create executor: %v", err)
	// }
	//
	// fromAddress := common.HexToAddress(req.FromAddress)
	// tx, err := executor.BuildEIP1559Transaction(ctx, quote.Tx, fromAddress)
	// if err != nil {
	//     log.Fatalf("Failed to build transaction: %v", err)
	// }
	//
	// txHash, err := executor.SignAndSendTransaction(ctx, tx, "YOUR_PRIVATE_KEY")
	// if err != nil {
	//     log.Fatalf("Failed to send transaction: %v", err)
	// }
	//
	// fmt.Printf("Transaction sent: %s\n", txHash.Hex())
	//
	// receipt, err := executor.WaitForTransaction(ctx, txHash)
	// if err != nil {
	//     log.Fatalf("Failed to wait for transaction: %v", err)
	// }
	//
	// fmt.Printf("Transaction confirmed in block: %d\n", receipt.BlockNumber)
}
*/
